src/EntityComponentManager.cc

```diff
@@ -853,6 +862,29 @@ void EntityComponentManager::AddEntityToMessage(msgs::SerializedStateMap &_msg,
     compIter->second.set_component(ostr.str());
 
     // TODO(anyone) Set component being removed once we have a way to queue it
+
+    // Add a component to the message and set it to be removed if the component
+    // exists in the removedComponents map.
+
+    bool entityHasRemovedComponent =
+      this->dataPtr->removedComponents.find(_entity) !=
+      this->dataPtr->removedComponents.end();
+    if (entityHasRemovedComponent)
+    {
+      auto removedComponent = this->dataPtr->removedComponents[_entity];
+      msgs::SerializedComponent cmp;
+
+      cmp.set_type(removedComponent.first);
+      cmp.set_remove(true);
+      // Empty data is needed for the component to be processed afterwards
+      cmp.set_component(" ");
+
+      (*(entIter->second.mutable_components()))[
+        static_cast<int64_t>(removedComponent.first)] = cmp;
+
+      // Remove component from map
+      this->dataPtr->removedComponents.erase(_entity);
```

Comment:
We should make it possible to clear `removedComponents` outside of this function. For example, `GuiRunner` calls `SetState` to synchronize its ECM and ends up calling `RemoveComponent`, but since it doesn't publish its ECM, `removedComponents` is never cleared and would grow indefinitely. I suggest creating a function `ClearRemovedComponents` and whoever uses the ECM is responsible for calling it after a time step or a sync, just like `ClearNewlyCreatedEntities` and `ProcessRemoveEntityRequests`. 

https://github.com/ignitionrobotics/ign-gazebo/blob/139020c8ec0af219efd848b84b4954c2c65936e9/src/SimulationRunner.cc#L713-L716
https://github.com/ignitionrobotics/ign-gazebo/blob/139020c8ec0af219efd848b84b4954c2c65936e9/src/gui/GuiRunner.cc#L103-L104

Reply:
I assume that `this->dataPtr->removedComponents.erase(_entity)` wouldn't be needed there, even if the ECM is actually publishing the message? I mean, it would be the responsibility of the server to call that method?

Reply:
Well, remember that the server and the GUI have their own instances of the `EntityComponentManager` (ECM) class. What the GUI does on its own ECM doesn't affect the server. In the future, we will probably add a way for the GUI's ECM to be sent to the server, but that's not the case right now.

When the GUI calls `RemoveComponent` on its instance of ECM,  the removed component is added to `removeComponents` of the GUI ECM. But since the GUI doesn't call `AddEntityToMessage` (because it doesn't need to publish its state),  the `removedComponents` container inside the GUI ECM will not be cleared.

It would be better if `removedComponents` gets cleared every time step regardless of whether `AddEntityToMessage` is called.

Reply:
I think I didn't explain myself very well.

I've added the `ClearRemovedComponents` function in 7992720, but I'm not sure if [this](https://github.com/ignitionrobotics/ign-gazebo/blob/79927200d07cd5bc5953dde14ee15182988b06ff/src/EntityComponentManager.cc#L796) is still necessary or it's enough if `SimulationRunner.cc` and `GuiRunner.cc` call `ClearRemovedComponents`?

Reply:
Ah, I see. Yeah, I don't think its necessary. In fact, I think we should remove it because leaving it in would make it so that only one caller of `EntityComponentManager::State` would get the list of removed components. The second caller won't get the removed components because `removedComponents` would have been cleared by the first call.